e2store/src/e2store/stream.rs

```diff
@@ -0,0 +1,89 @@
+use std::{
+    fs::File,
+    io::{Read, Write},
+    path::PathBuf,
+};
+
+use super::types::{Entry, Header};
+
+/// e2s.rs was built to load full .era/.era2 files into memory and provide a simple API to access
+/// the data. The issue for this is for larger files this wouldn't be feasible, as the entire file
+/// would need to be loaded into memory. This is where e2store_file.rs comes in, it provides a way
+/// to read and write e2store files in a streaming fashion.
+pub struct E2StoreStream {
+    pub e2store_file: File,
+}
+
+impl E2StoreStream {
```

Comment:
I would say it makes more sense to have 2 separate structures: `E2StoreStreamReader` and `E2StoreStreamWriter` (one that can only read and one that can only write).

Reply:
I like https://github.com/ethereum/trin/pull/1373#discussion_r1714958390 suggestion

Ideally we match the API with the same API as something like File provides in std, obviously it will be different, but you get what I mean. I think it is consistent with already popular paradigms which I think is good

Reply:
I was going with the direction of `BufReader` and `BufWriter`, which is also popular std paradigm. Even the documentation of the [File](https://doc.rust-lang.org/std/fs/struct.File.html) says:

_File does not buffer reads and writes. For efficiency, consider wrapping the file in a BufReader or BufWriter when performing many small read or write calls, unless unbuffered reads and writes are required._

---

And currently, we use mix of `write` and `deserialize` functions which isn't consistent.
If I would refactor it, I would name them `read` and `write` and they would accept `impl Read` and `impl Write` objects (instead of byte slices).